api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+						continue
+					}
+
+					usedPorts = append(usedPorts, port)
+				}
+			}
+		}
+
+		inUse := aws.Int64Value(instance.PendingTasksCount)+aws.Int64Value(instance.RunningTasksCount) > 0
+		r := &models.ResourceProvider{
+			ID:              aws.StringValue(instance.Ec2InstanceId),
+			InUse:           inUse,
+			UsedPorts:       usedPorts,
+			AvailableMemory: fmt.Sprintf("%v", availableMemory), //todo: this is definitely not correct
+		}
+
+		result = append(result, r)
+	}
+
+	return result, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers(clusterName string) ([]models.ResourceConsumer, error) {
+	// from scaler in develop, there are funcs to aggregate three types of resources
+	//   - getPendingServiceResources
+	//   - getPendingTaskResourcesInECS
+	//   - getPendingTaskResourcesInJobs
+
+	// GET PENDING SERVICE RESOURCE CONSUMERS
+	listServicesInput := &ecs.ListServicesInput{}
+	listServicesInput.SetCluster(clusterName)
+
+	serviceARNs := []*string{}
+	listServicesPagesFN := func(output *ecs.ListServicesOutput, lastPage bool) bool {
+		serviceARNs = append(serviceARNs, output.ServiceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListServicesPages(listServicesInput, listServicesPagesFN); err != nil {
+		return nil, err
+	}
+
+	services := []*ecs.Service{}
+	if len(serviceARNs) > 0 {
+		// The SDK states that you can specify up to 10 services in one DescribeServices operation:
+		// https://github.com/aws/aws-sdk-go/blob/ee1f179877b2daf2aaabf71fa900773bf8842253/service/ecs/api.go#L5420
+		// (aws-sdk-go version 1.12.19, as stated in layer0/Gopkg.toml)
+		for i := 0; i < len(serviceARNs); i += 10 {
+			end := i + 10
+
+			if end > len(serviceARNs) {
+				end = len(serviceARNs)
+			}
+
+			describeServicesInput := &ecs.DescribeServicesInput{}
+			describeServicesInput.SetCluster(clusterName)
+			describeServicesInput.SetServices(serviceARNs[i:end])
+
+			output, err := e.Client.ECS.DescribeServices(describeServicesInput)
+			if err != nil {
+				return nil, err
+			}
+
+			services = append(services, output.Services...)
+		}
+	}
+
+	resourceConsumers := []models.ResourceConsumer{}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int64{}
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
+			runningCount := aws.Int64Value(d.RunningCount)
+			pendingCount := aws.Int64Value(d.PendingCount)
+			if numPending := desiredCount - (runningCount + pendingCount); numPending > 0 {
+				deployIDCopies[aws.StringValue(d.Id)] = numPending
+			}
+		}
+
+		if len(deployIDCopies) == 0 {
+			continue
+		}
+
+		// iterate through deploys
+		for deployID := range deployIDCopies {
+			c, err := e.getContainerResourceFromDeploy(deployID)
+			if err != nil {
+				return nil, err
+			}
+
+			resourceConsumers = append(resourceConsumers, c...)
+		}
+	}
+
+	// GET PENDING TASK RESOURCE CONSUMERS IN ECS
+	taskSummaries, err := e.TaskProvider.List()
```

Comment:
❗️ This is super inefficient. Services had a similar pattern (in develop) where all entities are querie, cluster by cluster for all the clusters. Then you are throwing away most of the results by filtering for a single environment cluster.

Lets only query the tasks from the environment that we're currently scaling.

Reply:
Would the aws api be used directly to filter input for single env

Reply:
Yup exactly that. If we look in the implementation for `e.TaskProvivider.List()`, it does an equivalent of an `environment.List()` and then calls a helper method called `listClusterTaskARNs` for each cluster/env. I think either exposing the listClusterTaskARNs to the provider somehow or duplicate the code in the scaler would be good approach (probably the latter might be simple to start with).